File: backend/engine/pacer.py
"""
HumanPacer — organic, randomized delays to mimic human browsing.

Uses weighted probability distribution:
- 70% short pauses (quick successive actions)
- 20% medium pauses (reading/thinking)
- 10% long pauses (distracted, phone down)
- Session fatigue: after a burst of actions, takes a big break
"""

import asyncio
import random

from backend.core.logger import log


class HumanPacer:
    def __init__(self, base_min: float = 4, base_max: float = 8) -> None:
        self.base_min = base_min
        self.base_max = base_max
        self._action_count = 0
        self._session_length = random.randint(12, 25)

    @property
    def action_count(self) -> int:
        return self._action_count

    def next_pause(self) -> float:
        """Return next pause duration for clean/heavy actions."""
        self._action_count += 1

        # Session fatigue: after a burst, big break (90-300s)
        if self._action_count >= self._session_length:
            self._action_count = 0
            self._session_length = random.randint(10, 25)
            delay = random.uniform(90, 300)
            log.info("pacer", f"Session break: {delay:.0f}s")
            return delay

        roll = random.random()
        if roll < 0.70:
            # Short pause: quick action
            return random.uniform(self.base_min * 0.5, self.base_max * 0.7)
        elif roll < 0.90:
            # Medium pause: reading a profile
            return random.uniform(self.base_max * 0.8, self.base_max * 1.8)
        else:
            # Long pause: human distraction
            return random.uniform(self.base_max * 2, self.base_max * 4)

    def next_scan_pause(self) -> float:
        """8-15s base pauses for scanning, with human variation."""
        self._action_count += 1

        if self._action_count >= self._session_length:
            self._action_count = 0
            self._session_length = random.randint(8, 18)
            delay = random.uniform(120, 240)
            log.info("pacer", f"Scan session break: {delay:.0f}s")
            return delay

        roll = random.random()
        if roll < 0.65:
            # Normal pace: 8-12s
            return random.uniform(8, 12)
        elif roll < 0.88:
            # Slower: reading the profile (12-18s)
            return random.uniform(12, 18)
        else:
            # Human distraction (20-35s)
            return random.uniform(20, 35)


async def isleep(seconds: float, stop_event: asyncio.Event, step: float = 0.2) -> None:
    """Interruptible async sleep — returns early if stop_event is set."""
    elapsed = 0.0
    while elapsed < seconds:
        if stop_event.is_set():
            return
        chunk = min(step, seconds - elapsed)
        await asyncio.sleep(chunk)
        elapsed += chunk

File: src/background/pacer.ts
/**
 * HumanPacer — organic, randomized delays to mimic human browsing.
 *
 * Ported from backend/engine/pacer.py with identical logic.
 */

function randomBetween(min: number, max: number): number {
  return min + Math.random() * (max - min);
}

export class HumanPacer {
  private baseMin: number;
  private baseMax: number;
  private _actionCount = 0;
  private _sessionLength: number;

  constructor(baseMin = 4, baseMax = 8) {
    this.baseMin = baseMin;
    this.baseMax = baseMax;
    this._sessionLength = Math.floor(randomBetween(12, 25));
  }

  get actionCount(): number {
    return this._actionCount;
  }

  /** Return next pause duration for clean/heavy actions. */
  nextPause(): number {
    this._actionCount++;

    // Session fatigue: after a burst, big break (90-300s)
    if (this._actionCount >= this._sessionLength) {
      this._actionCount = 0;
      this._sessionLength = Math.floor(randomBetween(10, 25));
      return randomBetween(90, 300);
    }

    const roll = Math.random();
    if (roll < 0.7) {
      // Short pause: quick action
      return randomBetween(this.baseMin * 0.5, this.baseMax * 0.7);
    } else if (roll < 0.9) {
      // Medium pause: reading a profile
      return randomBetween(this.baseMax * 0.8, this.baseMax * 1.8);
    } else {
      // Long pause: human distraction
      return randomBetween(this.baseMax * 2, this.baseMax * 4);
    }
  }

  /** 8-15s base pauses for scanning, with human variation. */
  nextScanPause(): number {
    this._actionCount++;

    if (this._actionCount >= this._sessionLength) {
      this._actionCount = 0;
      this._sessionLength = Math.floor(randomBetween(8, 18));
      return randomBetween(120, 240);
    }

    const roll = Math.random();
    if (roll < 0.65) {
      return randomBetween(8, 12);
    } else if (roll < 0.88) {
      return randomBetween(12, 18);
    } else {
      return randomBetween(20, 35);
    }
  }
}

/** Sleep that can be cancelled via AbortSignal. */
export function sleep(seconds: number, signal?: AbortSignal): Promise<void> {
  return new Promise((resolve) => {
    if (signal?.aborted) {
      resolve();
      return;
    }
    const id = setTimeout(resolve, seconds * 1000);
    signal?.addEventListener("abort", () => {
      clearTimeout(id);
      resolve();
    });
  });
}
